src/test/java/edu/wisc/my/messages/model/ValidDateTest.java

```diff
@@ -13,31 +14,144 @@ public void filterOutExpiredDates() {
+      String longFutureDate = "2999-12-31";
+      message.setExpireDate(longFutureDate);
+      assertTrue("Future expiration date omitting time should not invalidate message",
+        message.isValidToday());
     }
 
     @Test
     public void filterOutNotYetLiveDates() {
         Message message = new Message();
         String futureDate = "2100-12-31";
         message.setGoLiveDate(futureDate);
-        assertFalse(message.isValidToday());
+        assertFalse("Future go-live date without tim specified should invalidate message.",
+          message.isValidToday());
+    }
+
+    @Test
+    public void pastGoLiveDatesYieldValidMessages() {
+      Message message = new Message();
+      String pastDate = "2000-04-12T12:21:21";
+      message.setGoLiveDate(pastDate);
+      assertTrue("Messages with past go-live date-times with time specified should be valid.",
+        message.isValidToday());
+    }
+
+    @Test
+    public void pastGoLiveDatesCanOmitTime() {
+      Message message = new Message();
+      String pastDate = "2000-04-12";
+      message.setGoLiveDate(pastDate);
+      assertTrue("Past go-live date omitting time should not invalidate message.",
+        message.isValidToday());
     }
 
     @Test
     public void filterOutImproperDates() {
         Message message = new Message();
         String nonsenseDate = "Not a date";
         message.setGoLiveDate(nonsenseDate);
-        assertFalse(message.isValidToday());
+        assertFalse("Unparseable date strings should invalidate a message (fail closed).",
+          message.isValidToday());
     }
 
     @Test
     public void nullDatesPass() {
         Message message = new Message();
         message.setExpireDate(null);
         message.setGoLiveDate(null);
-        assertTrue(message.isValidToday());
+        assertTrue("Null go live or expire dates should not invalidate message.",
+          message.isValidToday());
+    }
+
+    @Test
+    public void ignoresEmptyStringDates() {
+      Message message = new Message();
+      message.setExpireDate("");
+      message.setGoLiveDate("");
+      assertTrue("Empty date strings shouldn't invalidate a message",
+        message.isValidToday());
     }
 
+  /**
+   * Test that messages that expire later today are not considered expired.
+   * That is, that expiration supports the THH:MM suffix on ISO date-times.
+   */
+  @Test
+  public void expiringLaterTodayIsNotExpired() throws InterruptedException {
+
+    // assumption: this test will not take more than 2 seconds.
+
+    // this test relies on two seconds from now being in the same date, so that an implementation
+    // that respects time will notice that the that timestamp is not yet expired whereas an
+    // implementation that only respects dates will consider that timestamp expired.
+    //
+    // therefore edge case: it is less than four seconds before the date rolls over
+    // in this case wait four seconds to escape the edge case
```

Comment:
Could this edge case be resolved by making `isValidToday` a pure function?
Passing `now` as a parameter that defaults to `LocalDateTime.now();`, but could be overwritten for testing purposes.

Reply:
Yes, refactoring to do the date String evaluation in dedicated `Predicate` objects that take the `LocalDateTime` to evaluate against at construction rather than assuming it's now makes this more testable. I suspect `isValidToday()` can be refactored away entirely in favor of a `MessageValidAtTime` predicate that also takes the time against which to evaluate at construction, making it eminently testable.